fdmi/plugins/motr-fdmi-app.md

```diff
@@ -1,33 +1,88 @@
+For each record, the FDMI Application or Plugin performs actions and sends acknowledgements back to the source instances using Motr RPC calls.
 
-Example of such record in /etc/motr/confd.xc file will be:
+## How to create and FDMI Application and run with examples
 
-' {0x6c| ((^l|1:81), 2, ^l|1:81, "", ^n|1:3, ^v|1:66, [1: "Bucket-Name"], [1: "192.168.12.2@tcp:12345:4:1"])},'
+All FDMI related service code and examples are in the `cortx-motr/fdmi` directory. 
 
-More details about the FDMI design and settings can be found in this link: 
+One of the FDMI Application example is within the `plugins` directory named as fdmi_sample_plugin. 
+
+This application is written in C using `motr/client.h` interface and it adds the new MOTR client to the cluster and starts listening for FDMI records or events that match the FDMI filter substrings.  
+
+The executable will be compiled as part of the initial Motr compilation using the Makefile.am dependency and the Makefile.sub file.
+
+This program then communicates with the fdmi_app python wrapper by printing to stdout the records. 
+
+To test this application we need to run the fdmi_app wrapper typing in the console the following command:
+
+`sudo ./fdmi_app`
+
+The basic arguments needed will be picked by default from the `etc/motr/confd.xc` config file if not specified at the time of running.
+
+We will need to have a CORTX-Motr cluster already up and running with FDMI capabilities by adding the fdmi_filters section in the CDF.yaml config file that it is used for deploying with `hctl` command.
+ 
+You can check if your cluster is running with FDMI filter capability by looking at the /etc/motr/confd.xc file in which case will have the specific filter-id specified. 
+
+A second sample test that also runs the fdmi_sample_plugin program can be showed by using the following command:
+
+`./s/fdmi_plugin_st/fdmi_plugin_st.sh`
+
+This shell script starts Motr services using a specific filter configuration in which case we do not need to deploy the CORTX-Motr cluster beforehand using `hctl` command and specifying the fdmi_filter subsection in the CDF.yaml config file.
+
+However, we need to modify manually the `/m0t1fs/linux_kernel/st/m0t1fs_common_inc.sh` withing the `build_conf()` function to add the FDMI_FILTER_SUBSTRINGS and the FDMI_FILTER to include the number of substrings used.
+
+The `fdmi_plugin_st.sh` script starts the fdmi_plugin_sample application to listen for specific records that match with the filter already setup and do some key-value operations that reflects this specifications. 
 
-[FDMI Design](https://github.com/Seagate/cortx-motr/blob/main/fdmi/fdmi.c).
+The output will show whether the test was successful or not.
 
+## Word count application for Hackathon event
+
+Another fdmi example is the modification of the `fdmi_app` python script, `fdmi_app_word_count` that prints to stdout the most popular words in new PUT object requests.
```

Comment:
I think we just need something like the following:  

1.  First follow the motr QSG guide to build motr.
2.  Then follow the hare QSG guide to get a motr cluster up and running.
    1.  You will need to edit the CDG to create a filter.
    2.  The filter should be just matching a string
    3.  Ideally match any key-value pair where "hello" shows up anywhere in the value
    4.  It might have to match the whole string; can it do substring match?
    5.  Tell the reader that currently FDMI is only working for KV pairs and not objects
3.  Then launch the hello\_world\_fdmi program which registers itself as a listener for that filter
    1.  Launch it in one terminal window and leave it running there
4.  Then explain how to use m0kv to put new KV pairs
    1.  Provide two examples: one which matches the filter and one which doesn't
    2.  Then tell the reader to use the two examples and watch the output of the hello\_world\_fdmi program
5.  The hello\_word\_fdmi program should say something like:
    1.  Match on KV pair key='whatever it is'
    2.  It should then actually read the value and count the number of times the word hello is in the value and print that

Then, a super awesome ambitious further example would be:

1.  Get the full stack running with S3
2.  Have the listener/filter somehow detect when new objects are created matching some criteria
    1.  Maybe having bucket B as a parent (might be possible)
    2.  Maybe having some substring in the name of the S3 object (might be possible)
    3.  Maybe having some substring in the body of the S3 object (probably not possible)
3.  Have the listener do word\_count on the new object
4.  Give some sample PUT commands which some sample text files

Reply:
I added steps for setting the CORTX-Motr cluster and deploying with the corresponding link in the cortx-hare repo and also explanation of corresponding CDF file configuration with the corresponding link example also in the cortx-hare repo.

Reply:
The first hello example is already added in a new section called Create and Test a Hello FDMI sample application. 
Second example including using S3 server is added as part of the fdmi_app_word_count script in the Word count application for Hackathon event using S3 PUT requests section.